### PR TITLE
Ua

### PR DESCRIPTION
**Pull request by rd2**
99% progress status for UA' feedback & reports (e.g. runner, JSON & MD reports). May still need some minor tweaking for a few labels in the reports. Probably a merge with master (and v2.2. release) in a couple of days.

### CODE REVIEW

File: json/midrise.json [hunk 1]
```diff
@@ -1,5 +1,5 @@
 {
-  "schema": "https://github.com/rd2/tbd/blob/master/tbd.schema.json",
+  "schema": "https://github.com/rd2/tbd/blob/ua/tbd.schema.json",
```
Comment (rd2) [inline json/midrise.json:2]: All example JSON files will be updated, with the merge with master.

File: lib/measures/tbd/measure.rb [hunk 2]
```diff
@@ -46,26 +51,61 @@ def exitTBD(runner, out = false, io = nil, surfaces = nil)
...
+  end
+
+  unless TBD.fatal? || ua.nil? || ua.empty?
+    if ua.has_key?(:en)
+      if ua[:en].has_key?(:b1) || ua[:en].has_key?(:b2)
+        runner.registerInfo("-")
+        runner.registerInfo(ua[:model])
+        tbd_log[:ua] = {}
+        ud_md_en = ua_md(ua, :en)
+        ud_md_fr = ua_md(ua, :fr)
```
Comment (rd2) [inline lib/measures/tbd/measure.rb:82]: MD-formatted report content (French & English) is retrieved here , yet only written out to '/files' after the directory sorting is done.

File: lib/ua.rb [hunk 3]
```diff
@@ -0,0 +1,844 @@
+require "openstudio"
+
+# Set reference values for points, edges & surfaces (& subsurfaces) to
+# compute Quebec energy code (Section 3.3) UA' comparison (2021).
+#
+# @param [Hash] surfaces Preprocessed collection of TBD surfaces
+# @param [Hash] sets A model's PSI sets
+#
+# @return [Bool] Returns true if successful in generating UA' reference values
+def qc33(surfaces, sets)
```
Comment (rd2) [inline lib/ua.rb:10]: Basically 3 steps in UA calculations:
- set (reference) targets with respect to given code or standard (here, down to individual edge)
- analyse, then summarize (hash, minimally-formatted text ... useful for runner info & reports)
- generate reports (e.g. MD-formatted files)

File: lib/ua.rb [hunk 4]
```diff
@@ -0,0 +1,844 @@
...
+      when :grade
+        ua[:en][:b1][k] = "grade : #{str}"
+        ua[:fr][:b1][k] = "tracé : #{str}"
+      else
+        ua[:en][:b1][k] = "other : #{str}"
+        ua[:fr][:b1][k] = "autres : #{str}"
+      end
+    end
+
+    # Deterministic sorting
```
Comment (rd2) [inline lib/ua.rb:408]: This works, but the code could be more compact (e.g. loops)

File: tbd.schema.json [hunk 5]
```diff
@@ -315,6 +322,27 @@
         "surfaces"
       ]
     },
+    "SubSurface": {
+      "description": "Suburface(s) U-factors (optional)",
```
Comment (rd2) [inline tbd.schema.json:326]: This is strictly to alloy users to set NRFC-rated U-factors (without having to rely on an E+ run to retrieve them).

File: tbd.schema.json [hunk 6]
```diff
@@ -441,6 +469,106 @@
           "title": "Global state",
           "type": "string"
         },
+        "ua": {
+          "title": "UA' assessment",
+          "type": "object",
+          "properties": {
+            "bloc1": {
```
Comment (macumber) [inline tbd.schema.json:476]: What is `bloc1` vs `bloc2`?

File: tbd.schema.json [hunk 7]
```diff
@@ -315,6 +322,27 @@
...
+          "title": "Overall U-factor (optional)",
+          "type": "number",
+          "minimum": 0.01,
+          "maximum": 10.0
+        }
+      },
+      "additionalProperties": false,
+      "required": [
+        "id",
+        "usi"
```
Comment (macumber) [inline tbd.schema.json:343]: Is `usi` optional or required?  Title says optional.

File: tbd.schema.json [hunk 8]
```diff
@@ -441,6 +469,106 @@
...
+                },
+                "other": {
+                  "type": "string"
+                }
+              },
+              "additionalProperties": false,
+              "required": ["summary"]
+            },
+            "bloc2": {
+              "title": "heating setpoints >= 18°C",
```
Comment (macumber) [inline tbd.schema.json:524]: Typo? Should it be "bloc2: spaces/zones with HSTP < 18°C"?

File: tbd.schema.json [hunk 9]
```diff
@@ -441,6 +469,106 @@
...
+                "grade": {
+                  "type": "string"
+                },
+                "other": {
+                  "type": "string"
+                }
+              },
+              "additionalProperties": false,
+              "required": ["summary"]
+            }
```
Comment (macumber) [inline tbd.schema.json:569]: Want to add a `bloc3` just for completeness?

File: lib/ua.rb [hunk 10]
```diff
@@ -0,0 +1,844 @@
...
+    return report
+  end
+
+  if lang == :en && ua[:en].has_key?(:objective)
+    report << "# #{ua[:en][:objective]}   "
+    report << "   "
+  elsif lang == :fr && ua[:fr].has_key?(:objective)
+    report << "# #{ua[:fr][:objective]}   "
+    report << "   "
+  end
```
Comment (macumber) [inline lib/ua.rb:701]:
```suggestion
  if ua[lang] && ua[lang].has_key?(:objective)
    report << "# #{ua[lang][:objective]}   "
    report << "   "
  end
```

Comment (macumber) [reply lib/ua.rb:701]: Changes like this might make it a bit simpler

File: lib/ua.rb [hunk 11]
```diff
@@ -0,0 +1,844 @@
...
+    report << "##### SUMMARY   "
+    report << "   "
+    report << "#{ua[:model]}   "
+    report << "   "
+  elsif lang == :fr && ua.has_key?(:model)
+    report << "##### SOMMAIRE   "
+    report << "   "
+    report << "#{ua[:model]}   "
+    report << "   "
+  end
```
Comment (macumber) [inline lib/ua.rb:721]:
```suggestion
  strings = {}
  strings[:en] = {}
  strings[:fr] = {}
  strings[:en]["SUMMARY"] = "SUMMARY"
  strings[:fr]["SUMMARY"] = "SOMMAIRE "
  
  if ua.has_key?(:model)
    report << "##### #{strings[lang]["SUMMARY"]}"
    report << "   "
    report << "#{ua[:model]}   "
    report << "   "
  end
```

File: tbd.schema.json (hunk 7 above)
Comment (rd2) [reply tbd.schema.json:343]: Thanks for pointing that out. It's required - it's the only reason TBD has a subsurface entry. Will fix.

File: lib/measures/tbd/measure.rb [hunk 12]
```diff
@@ -46,26 +51,61 @@ def exitTBD(runner, out = false, io = nil, surfaces = nil)
...
   io = {} unless io
 
-  seed_file = runner.workflow.seedFile
-  seed_file = seed_file.get.to_s unless seed_file.empty?
-  description = "Thermal Bridging and Derating"
-  description += " - '#{seed_file}'" unless seed_file.empty?
-  io[:description] = description unless io.has_key?(:description)
+  seed = runner.workflow.seedFile
+  seed = seed.get.to_s unless seed.empty?
+  seed = File.basename(seed)
```
Comment (macumber) [inline lib/measures/tbd/measure.rb:56]:
```suggestion
  seed = File.basename(seed) unless seed.empty?
```

File: tbd.schema.json (hunk 8 above)
Comment (rd2) [reply tbd.schema.json:524]: That's a lazy copy/paste-driven error. Will fix.

File: tbd.schema.json (hunk 9 above)
Comment (rd2) [reply tbd.schema.json:569]: For the moment, I only have 2x categories to consider: "above 18°C" vs "below 18°C". If ever (in the future) a need for a 3rd (or 4th or nth) option pops up, I think I would revise the strategy and switch over to an (arbitrarily-sized) array of blocks (or bins).

File: lib/measures/tbd/measure.rb [hunk 13]
```diff
@@ -186,15 +247,27 @@ def arguments(model)
     write_tbd_json.setDefaultValue(false)
     args << write_tbd_json
 
+    gen_UA_report = OpenStudio::Measure::OSArgument.makeBoolArgument("gen_UA_report", true, false)
+    gen_UA_report.setDisplayName("Generate UA' report")
+    gen_UA_report.setDescription("Compare ∑U•A + ∑PSI•L + ∑KHI•n (model vs UA' reference - see pull-down option below).")
```
Comment (macumber) [inline lib/measures/tbd/measure.rb:252]: The sum character might print as □ if this ever runs in a Ruby instance with a strict ASCII character set

**Pull request merged**